lib/enumerable_1.rb

```diff
@@ -0,0 +1,98 @@
+# frozen_string_literal: true
+
+module Enumerable
+  def my_each
+    # return an enumerator
+    return each unless block_given?
+
+    i = 0
+    while i < size
+      yield(self[i])
+      i += 1
+    end
+    self
+  end
+
+  def my_each_with_index
+    return my_each unless block_given?
+
+    i = 0
+    while i < size
+      yield(self[i], i)
+      i += 1
+    end
+    self
+  end
+
+  def my_select
+    return my_each unless block_given?
+
+    arr = []
+    my_each do |item|
+      arr.push(item) if yield(item)
+    end
+    arr
+  end
+
+  def my_all?(pattern = nil)
+    condition = if block_given?
+                  proc { |item, _patt| yield(item) }
+                elsif !pattern.nil?
+                  proc do |item, patt|
+                    if patt.is_a? Regexp
+                      patt.match?(item)
+                    else
+                      item.is_a? patt
+                    end
+                  end
+                else
+                  proc { |item, _patt| item }
+                end
+
+    my_each do |item|
+      return false unless condition.call(item, pattern)
+    end
+    true
+  end
+
+  def my_any?(pattern = nil)
+    condition = if block_given?
+                  proc { |item, _patt| yield(item) }
+                elsif !pattern.nil?
+                  proc do |item, patt|
+                    if patt.is_a? Regexp
+                      patt.match?(item)
+                    else
+                      item.is_a? patt
+                    end
+                  end
+                else
+                  proc { |item, _patt| item }
+                end
+    my_each do |item|
+      return true if condition.call(item, pattern)
+    end
+    false
+  end
+
+  def my_none?(pattern = nil)
+    condition = if block_given?
+                  proc { |item, _patt| yield(item) }
+                elsif !pattern.nil?
+                  proc do |item, patt|
+                    if patt.is_a? Regexp
+                      patt.match?(item)
+                    else
+                      item.is_a? patt
+                    end
+                  end
+                else
+                  proc { |item, _patt| item }
+                end
+
+    my_each do |item|
+      return false if condition.call(item, pattern)
+    end
+    true
+  end
+end
```

Comment:
- [ ] should be identical and returns the same thing as ruby's [`none?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-none-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_none.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [ ] when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern